magma/linking.py

```diff
@@ -0,0 +1,70 @@
+from typing import Iterable, Mapping, Optional
+
+from magma.circuit import CircuitKind
+
+
+class _LinkInfo:
+    def __init__(self):
+        self.default_linked_module = None
+        self.linked_modules = {}
+
+
+_link_map = {}
+
+
+def _get_or_create_link_info(source: CircuitKind) -> _LinkInfo:
+    global _link_map
+    return _link_map.setdefault(source, _LinkInfo())
+
+
+def link_module(source: CircuitKind, key: str, target: CircuitKind):
+    # TODO(rsetaluri): Catch errors early here?
+    info = _get_or_create_link_info(source)
+    info.linked_modules[key] = target
+
+
+def get_linked_modules(source: CircuitKind) -> Mapping[str, CircuitKind]:
+    info = _get_or_create_link_info(source)
+    return info.linked_modules
+
+
+def link_default_module(source: CircuitKind, target: CircuitKind):
+    # TODO(rsetaluri): Catch errors early here?
+    info = _get_or_create_link_info(source)
+    info.default_linked_module = target
```

Comment:
Maybe we allow "overriding" the default linked module?  We could issue a warning by default and allow the user to pass a flag to suppress it ("override_default=True")

Reply:
My only concern is that the rules basically will get duplicated on the front-end and backend. (In terms of checking default, checking linking against the same key, etc.). I think that kind of interface^ is reasonable, but I'd rather delay more logic in the front-end until this becomes more of a hardened feature -- especially if we're migrating backends.